File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        return self.dfsGoodNodes(root, root.val)

    def dfsGoodNodes(self, node: TreeNode, maxValue: int) -> int:
        if not node: return 0
        result = 1 if node.val >= maxValue else 0
        maxValue = max(node.val, maxValue)
        result += self.dfsGoodNodes(node.left, maxValue) + self.dfsGoodNodes(node.right, maxValue)
        return result

File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int goodNodes(TreeNode root) {
        return dfsGoodNodes(root, root.val);
    }

    public int dfsGoodNodes(TreeNode node, int maxValue){
        if (node == null) return 0;
        int result = node.val >= maxValue ? 1: 0;
        maxValue = Math.max(node.val, maxValue);
        result += dfsGoodNodes(node.left, maxValue) + dfsGoodNodes(node.right, maxValue);
        return result;
    }
}